internal/ordered/map.go

```diff
@@ -0,0 +1,446 @@
+// Package ordered implements an ordered map type.
+package ordered
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"github.com/buildkite/agent/v3/internal/yamltojson"
+	"github.com/google/go-cmp/cmp"
+	"gopkg.in/yaml.v3"
+)
+
+var _ interface {
+	json.Marshaler
+	json.Unmarshaler
+	yaml.IsZeroer
+	yaml.Marshaler
+	yaml.Unmarshaler
+} = (*Map[string, any])(nil)
+
+// Map is an order-preserving map with string keys. It is intended for working
+// with YAML in an order-preserving way (off-spec, strictly speaking) and JSON
+// (more of the same).
+type Map[K comparable, V any] struct {
+	items []Tuple[K, V]
+	index map[K]int
+}
+
+// MapSS is a convenience alias to reduce keyboard wear.
+type MapSS = Map[string, string]
+
+// MapSA is a convenience alias to reduce keyboard wear.
+type MapSA = Map[string, any]
+
+// NewMap returns a new empty map with a given initial capacity.
+func NewMap[K comparable, V any](cap int) *Map[K, V] {
+	return &Map[K, V]{
+		items: make([]Tuple[K, V], 0, cap),
+		index: make(map[K]int, cap),
+	}
+}
+
+// MapFromItems creates an Map with some items.
+func MapFromItems[K comparable, V any](ps ...Tuple[K, V]) *Map[K, V] {
+	m := NewMap[K, V](len(ps))
+	for _, p := range ps {
+		m.Set(p.Key, p.Value)
+	}
+	return m
+}
+
+// Len returns the number of items in the map.
+func (m *Map[K, V]) Len() int {
+	if m == nil {
+		return 0
+	}
+	return len(m.index)
+}
+
+// IsZero reports if m is nil or empty. It is used by yaml.v3 to check
+// emptiness.
+func (m *Map[K, V]) IsZero() bool {
+	return m == nil || len(m.index) == 0
+}
+
+// Get retrieves the value associated with a key, and reports if it was found.
+func (m *Map[K, V]) Get(k K) (V, bool) {
+	var zv V
+	if m == nil {
+		return zv, false
+	}
+	idx, ok := m.index[k]
+	if !ok {
+		return zv, false
+	}
+	return m.items[idx].Value, true
+}
+
+// Set sets the value for the given key. If the key exists, it remains in its
+// existing spot, otherwise it is added to the end of the map.
+func (m *Map[K, V]) Set(k K, v V) {
+	// Suppose someone makes Map with new(Map). The one thing we need to not be
+	// nil will be nil.
+	if m.index == nil {
+		m.index = make(map[K]int, 1)
+	}
+
+	// Replace existing value?
+	if idx, exists := m.index[k]; exists {
+		m.items[idx].Value = v
+		return
+	}
+
+	// Append new item.
+	m.index[k] = len(m.items)
+	m.items = append(m.items, Tuple[K, V]{
+		Key:   k,
+		Value: v,
+	})
+}
+
+// Replace replaces an old key in the same spot with a new key and value.
+// If the old key doesn't exist in the map, the item is inserted at the end.
+// If the new key already exists in the map (and isn't equal to the old key),
+// then it is deleted.
+func (m *Map[K, V]) Replace(old, new K, v V) {
```

Comment:
Is the main reason this exists to avoid an unnecessary compact and allocation, compared to deleting `old` first and setting new later? I'd advertise that you can do that in the doc comments.

Reply:
The other problem is that if you delete `old` and then set `new`, then `new` won't be in the place of `old` if it isn't already there. So it's an easy way to change a key in-place. I'll add a comment.